File: frontend/js/views/logs.js
Views.logs = {
  title: "Message Logs",
  state: { page: 1 },

  async render(container) {
    this.container = container;
    setTitle("Message Logs");
    const campaigns = await API.get("/api/campaigns");

    container.innerHTML = `
      <div class="toolbar">
        <input id="f-phone" placeholder="Phone number">
        <select id="f-campaign"><option value="">All campaigns</option>${campaigns.map((c) => `<option value="${c.id}">${escapeHtml(c.name)}</option>`).join("")}</select>
        <select id="f-status">
          <option value="">All statuses</option>
          ${["sent", "delivered", "read", "failed", "received"].map((s) => `<option value="${s}">${s}</option>`).join("")}
        </select>
        <input id="f-template" placeholder="Template name">
        <input id="f-from" type="date">
        <input id="f-to" type="date">
        <button class="btn" id="btn-filter-logs">Filter</button>
      </div>
      <div class="panel">
        <div class="table-wrap"><table>
          <thead><tr><th>Time</th><th>Contact</th><th>Phone</th><th>Direction</th><th>Message</th><th>Template</th><th>Campaign</th><th>Status</th></tr></thead>
          <tbody id="logs-tbody"><tr><td colspan="8" class="empty">Loading…</td></tr></tbody>
        </table></div>
      </div>
      <div id="logs-pagination" class="toolbar" style="margin-top:14px"></div>
    `;
    document.getElementById("btn-filter-logs").onclick = () => { this.state.page = 1; this.load(); };
    await this.load();
  },

  async load() {
    const params = new URLSearchParams({ page: this.state.page, limit: 50 });
    const phone = document.getElementById("f-phone").value.trim();
    const campaign_id = document.getElementById("f-campaign").value;
    const status = document.getElementById("f-status").value;
    const template_name = document.getElementById("f-template").value.trim();
    const dateFrom = document.getElementById("f-from").value;
    const dateTo = document.getElementById("f-to").value;
    if (phone) params.set("phone", phone);
    if (campaign_id) params.set("campaign_id", campaign_id);
    if (status) params.set("status", status);
    if (template_name) params.set("template_name", template_name);
    if (dateFrom) params.set("date_from", dateFrom);
    if (dateTo) params.set("date_to", dateTo + "T23:59:59");

    const res = await API.get(`/api/logs?${params}`);
    document.getElementById("logs-tbody").innerHTML = res.data.length ? res.data.map((m) => `
      <tr>
        <td>${formatDate(m.created_at)}</td>
        <td>${escapeHtml(m.contact_name || "—")}</td>
        <td>${escapeHtml(m.phone)}</td>
        <td>${m.direction === "in" ? "Received" : "Sent"}</td>
        <td>${escapeHtml((m.body || "").slice(0, 60))}</td>
        <td>${escapeHtml(m.template_name || "—")}</td>
        <td>${escapeHtml(m.campaign_name || "—")}</td>
        <td>${badge(m.status)}</td>
      </tr>`).join("") : `<tr><td colspan="8" class="empty">No messages match these filters</td></tr>`;

    const totalPages = Math.max(1, Math.ceil(res.total / res.limit));
    document.getElementById("logs-pagination").innerHTML = `
      <button class="btn secondary small" ${this.state.page <= 1 ? "disabled" : ""} id="logs-prev">← Prev</button>
      <span class="text-dim">Page ${this.state.page} of ${totalPages} (${res.total} messages)</span>
      <button class="btn secondary small" ${this.state.page >= totalPages ? "disabled" : ""} id="logs-next">Next →</button>
    `;
    const prev = document.getElementById("logs-prev"), next = document.getElementById("logs-next");
    if (prev) prev.onclick = () => { this.state.page--; this.load(); };
    if (next) next.onclick = () => { this.state.page++; this.load(); };
  },
};

File: backend/routers/logs.py
from fastapi import APIRouter
from database import db_cursor

router = APIRouter(prefix="/api/logs", tags=["logs"])


@router.get("")
def message_logs(
    campaign_id: int | None = None,
    contact_id: int | None = None,
    phone: str | None = None,
    status: str | None = None,
    template_name: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
    page: int = 1,
    limit: int = 100,
):
    where = []
    params: list = []
    if campaign_id:
        where.append("m.campaign_id = ?")
        params.append(campaign_id)
    if contact_id:
        where.append("m.contact_id = ?")
        params.append(contact_id)
    if phone:
        where.append("c.phone LIKE ?")
        params.append(f"%{phone}%")
    if status:
        where.append("m.status = ?")
        params.append(status)
    if template_name:
        where.append("m.template_name LIKE ?")
        params.append(f"%{template_name}%")
    if date_from:
        where.append("m.created_at >= ?")
        params.append(date_from)
    if date_to:
        where.append("m.created_at <= ?")
        params.append(date_to)

    where_sql = ("WHERE " + " AND ".join(where)) if where else ""

    with db_cursor() as cur:
        total = cur.execute(f"""
            SELECT COUNT(*) FROM messages m JOIN contacts c ON m.contact_id = c.id {where_sql}
        """, params).fetchone()[0]

        offset = (page - 1) * limit
        rows = cur.execute(f"""
            SELECT m.*, c.name AS contact_name, c.phone, camp.name AS campaign_name
            FROM messages m
            JOIN contacts c ON m.contact_id = c.id
            LEFT JOIN campaigns camp ON m.campaign_id = camp.id
            {where_sql}
            ORDER BY m.created_at DESC LIMIT ? OFFSET ?
        """, params + [limit, offset]).fetchall()

    return {"total": total, "page": page, "limit": limit, "data": [dict(r) for r in rows]}
